batchflow/utils_transforms.py

```diff
@@ -0,0 +1,179 @@
+""" Transformations of ranges of values: normalizations and quantizations. """
+import numpy as np
+
+class Normalizer:
+    """ Class to hold parameters and methods for (de)normalization with provided stats.
+    Depending on the parameters, stats for normalization will be taken from (in order of priority):
+        - supplied `normalization_stats`, if provided
+        - computed from `array` directly
+
+    Parameters
+    ----------
+    mode : {'mean', 'std', 'meanstd', 'minmax'} or callable
+        If str, then normalization description.
+        If callable, then it will be called on `src` data with additional `normalization_stats` argument.
+    clip_to_quantiles : bool
+        Whether to clip the data to quantiles, specified by `q` parameter.
+        Quantile values are taken from `normalization_stats`, provided by either of the ways.
+    q : tuple of numbers
+        Quantiles for clipping. Used as keys to `normalization_stats`, provided by either of the ways.
+    normalization_stats : dict, optional
+        If provided, then used to get statistics for normalization.
+    inplace : bool
+        Whether to apply operation inplace or return a new array.
+    """
+    def __init__(self, mode='meanstd', clip_to_quantiles=False, q=(0.01, 0.99),
+                 normalization_stats=None, inplace=False):
+        self.mode = mode
+        self.clip_to_quantiles = clip_to_quantiles
+        self.q = q
+        self.normalization_stats = normalization_stats
+        self.inplace = inplace
+
+    def normalize(self, array, normalization_stats=None, return_stats=False):
+        """ Normalize image with provided stats.
+
+        Parameters
+        ----------
+        array : numpy.ndarray
+
+        normalization_stats : dict or None, optional
+            If provided, then used to get statistics for normalization, by default None.
+            If None, self.normalization_stats will be used.
+            If self.normalization_stats is also None, then statistics will by computed by array.
+        return_stats : bool, optional
+            Whether to return stats used for normalization, by default False
+
+        Returns
+        -------
+        numpy.ndarray or (numpy.ndarray, dict)
+        """
+        clip_to_quantiles = self.clip_to_quantiles
+        array = array if self.inplace else array.copy()
+
+        if normalization_stats is None:
+            if self.normalization_stats is not None:
+                normalization_stats = self.normalization_stats
+            else:
+                if clip_to_quantiles:
+                    np.clip(array, *np.quantile(array, self.q), out=array)
+                    clip_to_quantiles = False
+
+                if callable(self.mode):
+                    normalization_stats = {
+                        'mean': np.mean(array),
+                        'std': np.std(array),
+                        'min': np.min(array),
+                        'max': np.max(array),
+                    }
```

Comment:
Do we even need those in case of callable `mode`? Maybe, pass an empty dict and let this callable mode do whatever